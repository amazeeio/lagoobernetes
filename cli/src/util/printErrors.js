// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { +message: string, +stack?: string };

export function printErrors(
  cerr: Cerr,
  ...errors: Array<LaguError | string>
): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(
      err =>
        typeof err === 'string'
          ? err
          : R.propOr(
            // ...otherwise, use the message property as a fallback
            R.prop('message', err),
            // Try to get the stack...
            'stack',
            err,
          ),
    ),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoobernetes.yml config file not found. Please create one with "lagoobernetes init".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoobernetes.yml config file with "lagoobernetes init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoobernetes login".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-login',
    });
  }
  const errorsPluralized = R.length(errors) === 1 ? 'an error' : 'errors';
  return printErrors(
    cerr,
    `Oops! The Lagoobernetes API returned ${errorsPluralized}:`,
    ...errors,
  );
}
