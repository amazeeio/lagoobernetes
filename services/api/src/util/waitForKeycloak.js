// @flow

const logger = require('../logger');
const { getKeycloakAdminClient } = require('../clients/keycloak-admin');

/* ::

type ConnectionSettings = {
  baseUrl?: string,
  realmName?: string,
  requestConfig?: Object,
}

type UserSettings = {
  username?: string,
  password?: string,
  grantType?: string,
  clientId?: string,
};

*/

async function waitForKeycloak() {
  let keycloakReady = false;
  let keycloakAdminClient;

  do {
    try {
      keycloakAdminClient = await getKeycloakAdminClient();

      if (!(await keycloakAdminClient.realms.findOne({ realm: 'lagoobernetes' }))) {
        throw new Error('The "lagoobernetes" realm has not been created yet.');
      }

      const clients = await keycloakAdminClient.clients.find({ clientId: 'api' });
      if (!clients.length) {
        throw new Error('The "api" client has not been created yet.');
      }

      keycloakReady = true;
    } catch (err) {
      logger.debug(`Waiting for Keycloak to start... (error was ${err})`);
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
  } while (!keycloakReady);

  if (!keycloakAdminClient) {
    throw new Error('Keycloak client not initialized!');
  }

  logger.debug('Connected to Keycloak');
}

module.exports = waitForKeycloak;
