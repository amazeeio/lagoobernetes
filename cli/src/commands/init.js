// @flow

import path from 'path';
import { green } from 'chalk';
import inquirer from 'inquirer';
import R from 'ramda';
import tildify from 'tildify';
import untildify from 'untildify';
import { answerWithOptionIfSetOrPrompt } from '../cli/answerWithOption';
import { globalOptionDefaults } from '../config/globalOptions';
import { writeConfigFile } from '../configFile/writeConfigFile';
import { fileExists } from '../util/fs';
import { printErrors } from '../util/printErrors';

import typeof Yargs from 'yargs';
import type { ConfigFileInput } from '../types/ConfigFile';
import type { CommandHandlerArgsWithOptions } from '../types/Command';

export const command = 'init';
export const description =
  'Create a .lagoobernetes.yml config file in the current working directory';

export const OVERWRITE: 'overwrite' = 'overwrite';
export const PROJECT: 'project' = 'project';
export const API: 'api' = 'api';
export const SSH: 'ssh' = 'ssh';
export const TOKEN: 'token' = 'token';

export const commandOptions = Object.freeze({
  [OVERWRITE]: OVERWRITE,
  [PROJECT]: PROJECT,
  [API]: API,
  [SSH]: SSH,
  [TOKEN]: TOKEN,
});

export const dynamicOptionsKeys = Object.freeze([
  OVERWRITE,
  PROJECT,
  API,
  SSH,
  TOKEN,
]);

type OptionalOptions = {
  overwrite?: boolean,
  project?: string,
  api?: string,
  ssh?: string,
  token: string,
};

export function builder(yargs: Yargs) {
  return yargs
    .usage(`$0 ${command} - ${description}`)
    .options({
      [OVERWRITE]: {
        describe: 'Overwrite the configuration file if it exists',
        type: 'boolean',
      },
      [PROJECT]: {
        describe: 'Name of project to configure',
        type: 'string',
      },
      [API]: {
        describe: 'API URL',
        type: 'string',
      },
      [SSH]: {
        describe: 'SSH URL',
        type: 'string',
      },
      [TOKEN]: {
        describe: 'Path to the Lagoobernetes token file',
        type: 'string',
      },
    })
    .example(
      `$0 ${command}`,
      'Create a config file at ./.lagoobernetes.yml. This will confirm with the user whether to overwrite the config if it already exists and also prompt for some parameters to add to the config.\n',
    )
    .example(
      `$0 ${command} --${OVERWRITE}`,
      'Overwrite existing config file (do not confirm with the user).\n',
    )
    .example(
      `$0 ${command} --${OVERWRITE} false`,
      'Prevent overwriting of existing config file (do not confirm with user).\n',
    )
    .example(
      `$0 ${command} --${PROJECT} my_project`,
      'Set project to "my_project" (do not prompt the user).\n',
    )
    .example(
      `$0 ${command} --${API} http://localhost:3000`,
      'Set API URL to "http://localhost:3000" (do not prompt the user).\n',
    )
    .example(
      `$0 ${command} --${SSH} localhost:2020`,
      'Set SSH URL to "localhost:2020" (do not prompt the user).\n',
    )
    .example(
      `$0 ${command} --${TOKEN} ~/tokens/.lagoobernetes-token`,
      'Set the token path to ~/tokens/.lagoobernetes-token (do not prompt the user).\n',
    )
    .example(
      `$0 ${command} --${API} --${SSH} --${TOKEN}`,
      'Skip configuration of API and SSH URLs and token path (do not prompt the user).\n',
    )
    .example(
      `$0 ${command} --${OVERWRITE} --${PROJECT} my_project --${API} --${SSH} --${TOKEN}`,
      'Overwrite existing config files, set project to "my_project" and skip configuration of API and SSH URLs and token path (do not confirm with or prompt the user for any parameters).',
    );
}

type PromptForOverwriteArgs = {|
  filepath: string,
  options: OptionalOptions,
  clog: typeof console.log,
|};

async function promptForOverwrite({
  filepath,
  options,
  clog,
}:
PromptForOverwriteArgs): Promise<{ [key: typeof OVERWRITE]: boolean }> {
  return inquirer.prompt([
    {
      type: 'confirm',
      name: OVERWRITE,
      message: `File '${filepath}' already exists! Overwrite?`,
      default: false,
      when: answerWithOptionIfSetOrPrompt({
        option: OVERWRITE,
        options,
        notify: true,
        clog,
      }),
    },
  ]);
}

type Args = CommandHandlerArgsWithOptions<{
  +overwrite?: boolean,
  +project?: string,
}>;

export async function handler({
  cwd,
  options,
  clog,
  cerr,
}:
Args): Promise<number> {
  const filepath = path.join(cwd, '.lagoobernetes.yml');
  const exists = await fileExists(filepath);

  const overwrite = !exists
    ? undefined
    : R.prop(
      OVERWRITE,
      await promptForOverwrite({
        filepath,
        options,
        clog,
      }),
    );

  if (exists && !overwrite) {
    return printErrors(cerr, {
      message: `Not overwriting existing file '${filepath}'.`,
    });
  }

  const defaultToken = R.prop(TOKEN, globalOptionDefaults);

  // $FlowFixMe inquirer$Answers is inexact, LagoobernetesConfigInput is exact
  const configInput: ConfigFileInput = await inquirer.prompt([
    {
      type: 'input',
      name: PROJECT,
      message: 'Enter the name of the project to configure.',
      validate: input => (input ? Boolean(input) : 'Please enter a project.'),
      when: answerWithOptionIfSetOrPrompt({
        option: PROJECT,
        options,
        notify: true,
        clog,
      }),
    },
    {
      type: 'input',
      name: API,
      message: 'Enter the API URL',
      when: answerWithOptionIfSetOrPrompt({
        option: API,
        options,
        notify: true,
        clog,
      }),
    },
    {
      type: 'input',
      name: SSH,
      message: 'Enter the SSH URL',
      when: answerWithOptionIfSetOrPrompt({
        option: SSH,
        options,
        notify: true,
        clog,
      }),
    },
    {
      type: 'input',
      name: TOKEN,
      message: `Change the path for the token (default: ${R.compose(
        tildify,
        R.prop(TOKEN),
      )(globalOptionDefaults)})`,
      when: answerWithOptionIfSetOrPrompt({
        option: TOKEN,
        options: (R.pickBy((val, key) =>
          // Filter to options that...
          R.or(
            // ...either don't have the key TOKEN...
            R.complement(R.equals)(TOKEN, key),
            // ...or those that are a TOKEN that are the default value.
            // This is to avoid writing the default value in the config file.
            R.complement(R.equals)(defaultToken, val),
          ),
        )(options): { ...typeof options }),
        notify: true,
        clog,
      }),
      filter: untildify,
    },
  ]);

  try {
    clog(`Creating file '${filepath}'...`);
    await writeConfigFile(filepath, configInput);
    clog(green('Configuration file created!'));
    return 0;
  } catch (e) {
    return printErrors(
      cerr,
      `Error occurred while creating config at ${filepath}:`,
      e,
    );
  }
}
