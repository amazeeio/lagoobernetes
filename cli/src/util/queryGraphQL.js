// @flow

import R from 'ramda';
import { getConfig } from '../config';
import { getApiConfig } from '../config/getApiConfig';
import { fileExists, readFile } from './fs';
import request from './request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function queryGraphQL({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = R.prop('token', getConfig());
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const { hostname, port } = getApiConfig();

  const protocol = R.equals(port, 443) ? 'https:' : 'http:';

  const options = {
    hostname,
    path: '/graphql',
    port,
    protocol,
    method: 'POST',
    headers,
    body: JSON.stringify(
      {
        query,
        variables,
      },
      null,
      pretty ? 2 : 0,
    ),
    rejectUnauthorized: false,
  };

  let response;

  try {
    response = await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always(
        `Could not connect to Lagoobernetes API at ${protocol}//${hostname}:${port}/graphql.`,
      ),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }

  return response;
}
