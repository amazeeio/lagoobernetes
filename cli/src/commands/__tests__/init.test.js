// @flow

import { fileExists } from '../../util/fs';
import { handler } from '../init';

jest.mock('fs');
jest.mock('../../config');
jest.mock('../../util/fs');

// Flow does not know which objects are actual mocks
// this function casts given parameter to JestMockFn
const _castMockForFlow = (mockFn: any): JestMockFn<any, any> => mockFn;

describe('handler', () => {
  // TODO: Also write a test where it bails
  it('should overwrite existing config file if overwrite option passed as true', async () => {
    _castMockForFlow(fileExists).mockImplementationOnce(() =>
      Promise.resolve(true),
    );

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await handler({
      clog,
      cerr,
      cwd,
      options: {
        format: 'table',
        overwrite: true,
        project: 'test_project',
        api: '',
        ssh: '',
        token: '',
      },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should not overwrite config file when `overwrite` option set to false', async () => {
    _castMockForFlow(fileExists).mockImplementationOnce(() =>
      Promise.resolve(true),
    );

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await handler({
      clog,
      cerr,
      cwd,
      options: {
        format: 'table',
        overwrite: false,
        project: 'test_project',
        api: '',
        ssh: '',
        token: '',
      },
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should write default yaml to given cwd + .lagoobernetes.yml', async () => {
    _castMockForFlow(fileExists).mockImplementationOnce(() =>
      Promise.resolve(false),
    );

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await handler({
      clog,
      cerr,
      cwd,
      options: {
        format: 'table',
        overwrite: false,
        project: 'test_project',
        api: '',
        ssh: '',
        token: '',
      },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
