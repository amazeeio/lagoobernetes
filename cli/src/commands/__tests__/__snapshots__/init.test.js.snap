// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler should not overwrite config file when \`overwrite\` option set to false 1`] = `
Array [
  Array [
    "[31mNot overwriting existing file 'some/path/.lagoobernetes.yml'.[39m",
  ],
]
`;

exports[`handler should overwrite existing config file if overwrite option passed as true 1`] = `
Array [
  Array [
    "[34m![39m Set \\"overwrite\\" option to \\"true\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"project\\" option to \\"test_project\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"api\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"ssh\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"token\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "Creating file 'some/path/.lagoobernetes.yml'...",
  ],
  Array [
    "[32mConfiguration file created![39m",
  ],
]
`;

exports[`handler should write default yaml to given cwd + .lagoobernetes.yml 1`] = `
Array [
  Array [
    "[34m![39m Set \\"project\\" option to \\"test_project\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"api\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"ssh\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"token\\" option to \\"\\" from arguments or config",
  ],
  Array [
    "Creating file 'some/path/.lagoobernetes.yml'...",
  ],
  Array [
    "[32mConfiguration file created![39m",
  ],
]
`;
